Save and restore the administrative status for Links

cmagnobarbosa opened the pull request:
Today, changes in the Link administrative status (enabled/disabled) are not stored
in the storehouse NApp. This commit solves this problem, ensuring persistence
for these modifications.

The restored of the saved data must be triggered manually using this REST Endpoint:

`GET http://0.0.0.0:8181/api/kytos/topology/v3/restore`

Fix https://github.com/kytos/topology/issues/119

cmagnobarbosa committed:
Add Link Persistence

Today, administrative changes the Link status, enabled/disabled, are not stored
in the NApp storehouse. This commit solves this problem, ensuring persistence
for these modifications.

ajoaoff commented on main.py:
```diff
@@ -115,32 +118,43 @@ def _restore_status(self, switches_status, interfaces_status):
-        interfaces_status = {}
         status = self.storehouse.get_data()
         if status:
             switches = status.get('network_status')['switches']
+            links = status.get('network_status')['links']
```
I didn't understand why you are using `get` for the first key. Network status not existing is an error, and should be raised. With this code, if it does not exist, a `TypeError` will be raised, instead of a `KeyError`, and `KeyError` is more meaninful.

ajoaoff commented on main.py:
```diff
@@ -115,32 +118,43 @@ def _restore_status(self, switches_status, interfaces_status):
+            links = status.get('network_status')['links']
+            # get link status
+            for link, link_attributes in links.items():
+                self.links_state[link] = link_attributes.get('enabled')
```
I believe link states should have booleans as values, so 
```suggestion
                self.links_state[link] = link_attributes.get('enabled', False)
```
to have a default value if `enabled` does not exist, or 
```suggestion
                self.links_state[link] = link_attributes['enabled']
```
to raise an error if `enabled` does not exist.

ajoaoff commented on main.py:
```diff
@@ -115,32 +118,43 @@ def _restore_status(self, switches_status, interfaces_status):
-                # get status the switches
-                switches_status[switch] = switch_attributes.get('enabled')
+                # get swicthes status
+                self.switches_state[switch] = switch_attributes.get('enabled')
```
```suggestion
                self.switches_state[switch] = switch_attributes['enabled']
```
otherwise `self.switches_state[switch]` will receive `None` if `switch_attributes` does not have key `enabled`.

cmagnobarbosa committed:
Now, when the `enabled` key is not found, an exception will be thrown

This prevents the restore module from retrieving inconsistent data stored in the
NApp storehouse.

Co-authored-by: Antonio Francisco <antonio@ansp.br>

ajoaoff commented on main.py:
```diff
@@ -153,9 +167,9 @@ def get_topology(self):
-            switches_status, interfaces_status = self._load_network_status()
-            self._restore_status(switches_status, interfaces_status)
-        except (KeyError, FileNotFoundError) as exc:
+            self._load_network_status()
+            self._restore_status()
+        except (TypeError, FileNotFoundError) as exc:
```
```suggestion
            self._load_network_status()
            self._restore_status()
        except (KeyError, FileNotFoundError) as exc:
```
You changed the behavior of `_load_network_status`. Now, if it fails, it will raise `KeyError`. As it was before, it could raise both `KeyError` or `TypeError`.

cmagnobarbosa replied:
Thanks, I made the requested change. Returning the handler to `KeyError`.
https://github.com/kytos/topology/blob/9d769df356b743e564780c648ac6945295cbbcf9/main.py#L172

cmagnobarbosa commented on main.py:
```diff
@@ -117,32 +120,45 @@ def _restore_status(self, switches_status, interfaces_status):
                          f'interface {interface_id} does not exist.')
                 raise KeyError(error)
-
+        # links
+        for link, state, in self.links_state.items():
+            try:
+                if state:
+                    self.links[link].enable()
+                else:
+                    self.links[link].disable()
+            except KeyError:
+                error = ('Error restoring link status.'
+                         f'The link {link} does not exist.')
+                raise KeyError(error)
         log.info('Network status restored.')
```
This logic is not working on the first run, because this code snippet requires the [add_links](https://github.com/kytos/topology/blob/6f0968b7c4c56ed5844522fdabac2cfe0f86aaaf/main.py#L576) method to fill in the self.links. Due to this problem, it was moved to the draft again.

cmagnobarbosa committed:
Create a Link before restoring the stored state

Link persistence now checks for a link before restoring the
link state stored in the NApp storehouse.

cmagnobarbosa commented on main.py:
```diff
@@ -88,10 +91,45 @@ def _get_link_from_interface(self, interface):
+                error_msg = (f"Error restoring link endpoint: {error}")
+                raise KeyError(error_msg)
+
+            link = self._get_link_or_create(endpoint_a, endpoint_b)
+            endpoint_a.update_link(link)
+            endpoint_b.update_link(link)
+
+            endpoint_a.nni = True
+            endpoint_b.nni = True
+
+            self.notify_topology_update()
```
These parts of the code create a new link with the data stored in the NApp storehouse.
